pytorch_lightning/plugins/training_type/ddp.py

```diff
@@ -304,6 +315,46 @@ def _register_ddp_hooks(self) -> None:
                 ddp_comm_wrapper=self._ddp_comm_wrapper,
             )
 
+            # Post-localSDG is only available after 1.9, and `torch.distributed.optim` package currently is not available on Windows.
+            if (
+                not _IS_WINDOWS
+                and _TORCH_GREATER_EQUAL_1_9
+                and isinstance(self._ddp_comm_state, post_localSGD.PostLocalSGDState)
+                and self.lightning_module.trainer.state.fn == TrainerFn.FITTING
+            ):
+                self._reinit_optimizers_with_post_localSGD(self._ddp_comm_state.start_localSGD_iter)
+
+    def _reinit_optimizers_with_post_localSGD(self, warmup_steps: int):
+        optimizers = self.lightning_module.trainer.optimizers
+        if self._model_averaging_period is None:
+            rank_zero_warn(
+                "Post-localSGD algorithm is used, but model averaging period is not provided to DDP plugin. Set this period as 16."
+            )
+            self._model_averaging_period = 16
+        averager = averagers.PeriodicModelAverager(period=self._model_averaging_period, warmup_steps=warmup_steps)
+        for x, optimizer in enumerate(optimizers):
+            if isinstance(optimizer, LightningOptimizer):
+                optimizer = optimizer._optimizer
+
+            if isinstance(optimizer, DistributedOptimizer) or isinstance(optimizer, ZeroRedundancyOptimizer):
+                raise ValueError(
+                    f"Cannot wrap a distributed optimizer of type {optimizer.__name__} by PostLocalSGDOptimizer."
+                )
+
+            if not isinstance(optimizer, PostLocalSGDOptimizer):
+                optim_class = type(optimizer)
+                post_localSGD_optimizer = PostLocalSGDOptimizer(
+                    params=self.model.parameters(),
```

Comment:
should we be re-initializing the parameters by looking up the param groups from the original optimizer? what happens if the lightning module contains multiple trainable modules? this would include all their parameters

reference from sharded data parallel with OSS: https://github.com/PyTorchLightning/pytorch-lightning/blob/5329b0d11358b72a42d596d3b70b7010e35c45af/pytorch_lightning/plugins/training_type/sharded.py#L54

Reply:
Replaced `self.model.parameters()` with `optimizer.param_groups`.